apps/tangle-dapp/data/blueprints/utils/toPrimitiveBlueprint.ts

```diff
@@ -0,0 +1,309 @@
+  return {
+    metadata: toPrimitiveJobMetadata(metadata),
+    params: params.map(toPrimitiveFieldType),
+    result: result.map(toPrimitiveFieldType),
+    verifier: toPrimitiveJobResultVerifier(verifier),
+  } as const;
+}
+
+export function toPrimitiveServiceRegistrationHook(
+  registrationHook: ServiceRegistrationHook,
+) {
+  if (registrationHook.type === 'Evm') {
+    return {
+      Evm: registrationHook.asEvm.toHex(),
+    } as const;
+  }
+
+  return registrationHook.type;
+}
+
+export function toPrimitiveServiceRequestHook(requestHook: ServiceRequestHook) {
+  if (requestHook.type === 'Evm') {
+    return {
+      Evm: requestHook.asEvm.toHex(),
+    } as const;
+  }
+
+  return requestHook.type;
+}
+
+export function toPrimitiveJobResultVerifier(verifier: JobResultVerifier) {
+  if (verifier.type === 'Evm') {
+    return {
+      Evm: verifier.asEvm.toHex(),
+    } as const;
+  }
+
+  return verifier.type;
+}
+
+export function toPrimitiveJobMetadata({ name, description }: JobMetadata) {
+  return {
+    name: u8aToString(name),
+    description: optionalBytesToString(description),
+  } as const;
+}
+
+export function toPrimitiveGadget(gadget: Gadget) {
+  if (gadget.type === 'Container') {
+    return toPrimitiveContainerGadget(gadget.asContainer);
+  }
+
+  if (gadget.type === 'Native') {
+    return toPrimitiveNativeGadget(gadget.asNative);
+  }
+
+  if (gadget.type === 'Wasm') {
+    return toPrimitiveWasmGadget(gadget.asWasm);
+  }
+
+  throw new Error('Unknown Gadget type');
+}
+
+export function toPrimitiveWasmGadget({ runtime, sources }: WasmGadget) {
+  return {
+    runtime: toPrimitiveWasmRuntime(runtime),
+    sources: sources.map(toPrimitiveGadgetSource),
+  } as const;
+}
+
+export function toPrimitiveWasmRuntime(runtime: WasmRuntime) {
+  return runtime.type;
+}
+
+export function toPrimitiveNativeGadget({ sources }: NativeGadget) {
+  return {
+    sources: sources.map(toPrimitiveGadgetSource),
+  } as const;
+}
+
+export function toPrimitiveContainerGadget({ sources }: ContainerGadget) {
+  return {
+    sources: sources.map(toPrimitiveGadgetSource),
+  } as const;
+}
+
+export function toPrimitiveGadgetSource(source: GadgetSource) {
+  return {
+    fetcher: toPrimitiveGadgetSourceFetcher(source.fetcher),
+  } as const;
+}
+
+export function toPrimitiveGadgetSourceFetcher(fetcher: GadgetSourceFetcher) {
+  if (fetcher.type === 'ContainerImage')
+    return {
+      ContainerImage: toPrimitiveContainerImage(fetcher.asContainerImage),
+    } as const;
+
+  if (fetcher.type === 'Github')
+    return {
+      Github: toPrimitiveGithubFetcher(fetcher.asGithub),
+    } as const;
+
+  if (fetcher.type === 'Ipfs')
+    return {
+      Ipfs: u8aToString(fetcher.asIpfs),
+    } as const;
+
+  if (fetcher.type === 'Testing')
+    return {
+      Testing: toPrimitiveTestingFetcher(fetcher.asTesting),
+    } as const;
+
+  throw new Error('Unknown GadgetSourceFetcher type');
+}
```

Comment:
`switch` might be more appropriate here, especially if you've covered all the possibilities of `fetcher.type`, since that would eliminate the need to throw a runtime error and instead rely on compile-type switch exhaustiveness from the type system